### PR TITLE
feat: parser

### PR DESCRIPTION
**Pull request by RLKRo**
# Description

This pr adds parser: a tool for conversions between different formats.

Class [`DFFProject`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L81) is used for conversions. It has methods [`from_python`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L380), [`to_python`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L406), [`from_yaml`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L459), [`to_yaml`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L463), [`from_graph`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L470), [`to_graph`](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/dff/utils/parser/dff_project.py#L474) to convert between different formats.

Examples of usage may be found [here](https://github.com/deeppavlov/dialog_flow_framework/tree/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/tests/parser/TEST_CASES).

### Path

An id of a node -- a sequence of node names that leads to the node. For example, here
```
a = {1: 2, 3: 4}
```
`2` has a path of `({module_name}, a, value, value_1)`. First element is the name of a namespace, second element is the name of an object, the third element specifies that an object is inside a value of an assignment, the fourth element specifies that an object is inside a value of a key-value pair where key is `1`.

`3` has a path of `({module_name}, a, value, key_3)`.

### Reference Object

An object that references another object. For example, here
```
import module
a = module.b
```
`module` is a reference object (it references another namespace). `module.b` is also a reference object (it references an object in another namespace). To resolve a reference object means to obtain an object it references.

### Script

A tuple of (a dictionary, a tuple of two strings and a tuple of two strings). The first element is a `script`, the second element is its `start_label`, the third element is its `fallback_label`.

### Script Initializer

An assignment of a call that initializes a script, e.g. `actor = Actor(script, start_label=start_label, fallback_label=fallback_label)` is a script initializer. Supported function calls: `Pipeline.from_script`, `Actor.__init__`. A script initializer is specified via `DFFProject.script_initializer` which is a ":"-separated string the first part of which is the name of a file which contains a call and the second part is the name of a target of an assignment.

For example, if a file `/script/main.py` consists of
```
from dff.pipeline import Pipeline as pl
from . import script, start_label, fallback_label

pipeline = pl.from_script(script, fallback_label=fallback_label, start_label=start_label)
```

project `DFFProject.from_python("/", "/script/main.py", script_initializer="pipeline")` or `DFFProject.from_python("/", "/script/main.py", script_initializer="script.main:pipeline")` would have `(script, start_label, fallback_label)` as script.

If `script_initializer` is not specified it is found automatically (provided namespaces contain one and only one `script_initializer`).

### 1. Python

This format is currently used in all examples -- a dff script is stored inside `.py` file as a dictionary. An example can be found [here](https://github.com/deeppavlov/dialog_flow_framework/tree/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/tests/parser/TEST_CASES/conversions/just_works/python_files).

### 2. Yaml

An example of this format can be found [here](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/tests/parser/TEST_CASES/conversions/just_works/script.yaml). The script is stored inside a `.yaml` file. The file stores a dictionary with the first-level key-value pair representing namespaces and objects in them -- a key is the name of a file, value is a dictionary of objects defined in that file. The name of the file is its dot-separated extension from `project_root_dir`. A dictionary of objects consists of pairs `object_name: object`.

**Note:** only objects that are used inside a script initializer are stored inside a script file.

### 3. Graph

An example of this format can be found [here](https://github.com/deeppavlov/dialog_flow_framework/blob/92619a0dfc19533ad017b9cb9a15b517f3aa0ae4/tests/parser/TEST_CASES/conversions/just_works/graph.json). This format is used to store a `networkx` graph. The graph has one attribute `graph` which contains the same dictionary that is stored in `.yaml` format.

Nodes of the graph are nodes in the script as well as some special nodes. Each node's id is a tuple of (strings or tuples of strings).

Script nodes' ids start with `"NODE"`. The second and the third elements are flow_name and node_name respectively. 

Script nodes have a `ref` attribute which stores a path to that node; and a `local` attribute which is set to `True` if a node is `dff.script.LOCAL`.

Special nodes that represent labels start with `"LABEL"`. The second element is the name of a lable, e.g. `"repeat"`. Other elements are tuples, where the first element is a name of an argument to a label call, and the second element is its value.

Unresolved nodes have an id `("NONE", )`.

Edges of the graph represent transitions. Beside `source` and `target` attributes they contain paths and string representations of label and condition of the transition.

# Checklist

- [x] I have covered the code with tests
- [ ] I have added comments to my code to help others understand it
- [ ] I have updated the documentation to reflect the changes
- [ ] I have performed a self-review of the changes

## Review ready status

- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/base_parser_object.py
- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/dff_project.py
- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/exceptions.py
- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/namespace.py
- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/utils.py
- [x] https://github.com/deeppavlov/dialog_flow_framework/blob/feat/parser/dff/utils/parser/yaml.py

### CODE REVIEW

**Review by kudep — CHANGES_REQUESTED** [review]
- [x] refactor naming (`engine_examples` ? maybe `core_examples` )

File: tests/parser/TEST_CASES/PYTHON3.8-/conversions/edge_cases/script.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  var: some_var
+  script:
+    "'start_flow'":
+      "'start_node'":
+        TRANSITIONS:
+          transitions['two']: '(len(transitions[1]) > 0 > (- 2) > (lambda : ((- 5) + var))() > func())'
+      "'label'":
+        TRANSITIONS: transitions[7]
+        MISC:
+          '1': "{\n    2: 7,\n}[{\n    3: 2,\n}[3]]"
```
Comment (kudep) [inline tests/parser/TEST_CASES/PYTHON3.8-/conversions/edge_cases/script.yaml:17]: So hard)

Comment (RLKRo) [reply tests/parser/TEST_CASES/PYTHON3.8-/conversions/edge_cases/script.yaml:17]: This is an example to test edge cases -- it's not supposed to showcase the parser (and will not become an example).

**Comment by avsakharov** [comment]
We should add a header and a description to
1) utils/parser/utils.py
2) utils/parser/exceptions
3) utils/parser/yaml

Or remove those from the documentation that you don't need there.

File: dff/utils/parser/base_parser_object.py [hunk 2]
```diff
@@ -0,0 +1,1325 @@
+"""
+Base Parser Objects
+--------------------
+This module defines parser objects -- nodes that form a tree.
```
Comment (avsakharov) [inline dff/utils/parser/base_parser_object.py:4]: Should we expand this description?

File: dff/utils/parser/base_parser_object.py [hunk 3]
```diff
@@ -0,0 +1,1325 @@
...
+from dff.utils.parser.exceptions import StarError, ParsingError
+from dff.utils.parser.utils import is_instance
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseParserObject(ABC):
+    """
+    An interface for other parser objects.
```
Comment (avsakharov) [inline dff/utils/parser/base_parser_object.py:54]: Should we expand the description in the classes?

File: dff/utils/parser/dff_project.py [hunk 4]
```diff
@@ -0,0 +1,864 @@
+"""
+DFF Project
+-----------
+This module defines a class that represents a DFF project --
+a collection of python source files that define a script and a Pipeline.
+
+Glossary
+--------
```
Comment (avsakharov) [inline dff/utils/parser/dff_project.py:8]: Should this headline be smaller than the previous one?

File: dff/utils/parser/dff_project.py [hunk 5]
```diff
@@ -0,0 +1,864 @@
...
+            Whether to perform validation -- check for a script initializer and validate its arguments.
+            Defaults to True.
+        :param script_initializer:
+            A colon-separated string that points to a script initializer call.
+            The first part of the string should be the name of the namespace.
+            The second part of the string should be the name of the object that is a result of script initializer call.
+            Instead of a colon-separated string a name of the object can be provided
+            in which case the name of the namespace is assumed to be the name of `entry_point`.
+            Defaults to None.
+        :return:
```
Comment (avsakharov) [inline dff/utils/parser/dff_project.py:690]: We should write something here, or remove it.

File: dff/utils/parser/namespace.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
+"""
+Namespace
+---------
+This module defines `Namespace` -- a class to store contents of python files.
```
Comment (avsakharov) [inline dff/utils/parser/namespace.py:4]: Should we add more specific description?

**Comment by avsakharov** [comment]
- [ ] Base Parser Objects: expand description?
- [ ] class BaseParserObject: expand description?
- [ ] Glossary: this headlaine should be smaller
- [ ] :return: write something or remove
- [ ] Namespace: more specific description?

**Review by avsakharov — APPROVED** [review]

File: tests/parser/parser_test_generators.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            raise RuntimeError(f"Main file {main_file} not found.")
+
+        dff_project = DFFProject.from_python(python_dir, main_file)
+
+        dff_project.to_yaml(working_dir / "script.yaml")
+        dff_project.to_graph(working_dir / "graph.json")
+
+        # Generate diff file
+
+        with open(working_dir / "script.yaml", "r") as fd:
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply tests/parser/parser_test_generators.py:56]: All files generated by the function are created under `TEST_CASES` and are supposed to be in the repository.

File: dff/utils/parser/base_parser_object.py [hunk 8]
```diff
@@ -0,0 +1,1312 @@
...
+    This class is for nodes that represent :py:class:`ast.stmt`.
+    """
+
+    def __init__(self):
+        BaseParserObject.__init__(self)
+        self.parent: tp.Optional[Namespace] = None
+        self.children: tp.MutableMapping[str, Expression] = {}
+
+    @classmethod
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply dff/utils/parser/base_parser_object.py]: Added for `abstractmethod` [here](https://github.com/deeppavlov/dialog_flow_framework/pull/59/commits/ac441630f0c01a0821aefdb816b6184aefc5c3d1).
Not required for `overload`:

> https://docs.python.org/3/library/typing.html?highlight=overload#typing.overload

> At runtime, calling a @overload-decorated function directly will raise [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError).

File: dff/utils/parser/base_parser_object.py [hunk 9]
```diff
@@ -0,0 +1,1312 @@
...
+            return None
+        for _cls in (Import, ImportFrom, Assignment, Python):
+            _cls = tp.cast(tp.Type[Statement], _cls)
+            obj = _cls.from_ast(node, **kwargs)
+            if obj is not None:
+                return obj
+        raise RuntimeError(node)
+
+
+class Expression(BaseParserObject, ABC):
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply dff/utils/parser/base_parser_object.py]: Split into `base_classes`, `expressions` and `statements` [here](https://github.com/deeppavlov/dialog_flow_framework/pull/59/commits/158817fac67c792b13dffd19a30ab3dd9a484316).
Now maximum file length is 700 (lines).

I may also consider splitting files further (e.g. by turning expressions into its own sub-package with 7 modules inside for each expression) but I'm not sure that is necessary.

File: setup.py [hunk 10]
```diff
@@ -80,6 +80,13 @@ def merge_req_lists(*req_lists: List[str]) -> List[str]:
     "pytelegrambotapi==4.5.1",
 ]
 
+parser_dependencies = [
+    "cached-property==1.5.2; python_version<'3.8'",
+    "astunparse==1.6.3; python_version<'3.9'",
+    "ruamel.yaml",
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply setup.py:86]: No, those libraries are not actively developed so I thought that it would be better to put constraints on them.

File: tests/parser/TEST_CASES/complex_cases/modular/python_files/main.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from script import pipeline
```
[the remark this answers is not in the document]
Comment (RLKRo) [reply tests/parser/TEST_CASES/complex_cases/modular/python_files/main.py]: Yes. The purpose of creating it is to simulate a real world example of DFF usage, where a pipeline is not defined in the main file but rather imported in it.

**Review by RLKRo — COMMENTED** [review]
> Should we consider renaming just_works into something else?

Renamed `just_works` to `simple_dff_project` and `modular` to `modular_dff_project` [here](https://github.com/deeppavlov/dialog_flow_framework/pull/59/commits/ec769c34ae1c6bc1c446dfe4bee069e157661ef4).

> Should we separate tests from tutorials?

Those are not really tests or tutorials. I created a [readme file](https://github.com/deeppavlov/dialog_flow_framework/pull/59/commits/2ef3b7d4f427b5a8a8a04d2f6e1d2284d86c790c) to explain those files:

> All these files serve two purposes:
> 1. Showcase parser and its capabilities.
> 2. Test against them to ensure that parser is working correctly.

Right now I'm torn between the two options (to store them inside tests or store them inside tutorials):

On one hand, they are used in testing and don't follow the usual structure of tutorials (py files with %-formatting). On the other hand, they show the capabilities of parser on real-life examples.

File: dff/utils/parser/utils.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+try:
+    from functools import cached_property
+except ImportError:
+    try:
+        from cached_property import cached_property  # type: ignore
+    except ImportError:
+        raise ModuleNotFoundError(
+            "Module `cached_property` is not installed. Install it with `pip install dff[parser]`."
+        )
+# todo: remove this when python3.7 support is dropped
```
Comment (pseusys) [inline dff/utils/parser/utils.py:19]: It is already dropped

**Comment by RLKRo** [comment]
Will be rewritten (again).
